tests/unit/Referrals/ReferralDetectorTest.php

```diff
@@ -0,0 +1,69 @@
+<?php
+namespace MailPoet\Referrals;
+
+use MailPoet\Settings\SettingsController;
+use MailPoet\WP\Functions as WPFunctions;
+use PHPUnit_Framework_MockObject_MockObject as MockObject;
+
+class ReferralDetectorTest extends \MailPoetUnitTest {
+
+  /** @var MockObject */
+  private $settings_mock;
+
+  /** @var MockObject */
+  private $wp_mock;
+
+  function _before() {
+    $this->settings_mock = $this->createMock(SettingsController::class);
+    $this->wp_mock = $this->createMock(WPFunctions::class);
+    if (!defined(ReferralDetector::REFERRAL_CONSTANT_NAME)) {
+      define(ReferralDetector::REFERRAL_CONSTANT_NAME, 'constant_referral_id');
+    }
+  }
+
+  function testItPrefersSettingsValueOverAll() {
+    $this->settings_mock
+      ->expects($this->once())
+      ->method('get')
+      ->willReturn('settings_referral_id');
+    $this->wp_mock
+      ->expects($this->never())
+      ->method('getOption');
+    $referral_detector = new ReferralDetector($this->wp_mock, $this->settings_mock);
+    expect($referral_detector->detect())->equals('settings_referral_id');
+  }
+
+  function testItPrefersOptionValueOverConstantAndStoreValueToSettings() {
```

Comment:
Again the name is wrong. Should be "and storeS" and "prefer to" is more common than "prefer over".